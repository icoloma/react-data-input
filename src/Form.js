import React from "react";
import PropTypes from "prop-types";

// Form is the (optional) container of BoundComponent instances.
export default class Form extends React.Component {
  constructor() {
    super();
    this.onSubmit = this.onSubmit.bind(this);
    this.validationComponents = [];
    this.ref = React.createRef();
  }

  getChildContext() {
    return {
      form: this
    };
  }

  addValidationComponent(c) {
    this.validationComponents.push(c);
  }

  removeValidationComponent(c) {
    const vc = this.validationComponents;
    const index = vc.indexOf(c);
    index !== -1 && vc.splice(index, 1);
  }

  // validate all child components
  // @return a Promise that resolves to true if validation passes for all fields, false otherwise
  validate() {
    const promises = this.validationComponents.map(c => c.validate());
    return Promise.all(promises).then(errorMessages => {
      return errorMessages.every(errorMessage => !errorMessage);
    });
  }

  // if all validations pass, invokes the onSubmit callback
  // @return a Promise to be invoked after the validation is done
  onSubmit(e) {
    e.preventDefault();
    return this.validate().then(results => {
      if (results) {
        this.props.onSubmit(e);
      } else {
        const firstInvalidElement = this.ref.current.querySelector(
          "[aria-invalid=true]"
        );
        firstInvalidElement && firstInvalidElement.focus();
      }
    });
  }

  render() {
    const { state, ...props } = this.props;
    return (
      <form {...props} onSubmit={this.onSubmit} noValidate ref={this.ref}>
        {this.props.children}
      </form>
    );
  }
}

Form.propTypes = {
  // the callback that will be invoked on submit, if validation passes
  onSubmit: PropTypes.func.isRequired,

  // the state that will be injected into nested elements
  state: PropTypes.object
};

Form.childContextTypes = {
  form: PropTypes.object
};
