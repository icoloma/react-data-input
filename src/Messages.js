import { escape } from "lodash";

// replaces messages. We cannot use ES6 templates (lack of context),
// and lodash does not pass CSP
function template(message, data) {
  var templateReplace = function(s, match) {
    return data[match.trim()];
  };

  return message.replace(/\$\{([^}]+)}/g, function(s, match) {
    return escape(templateReplace(s, match));
  });
}

let contents = {
  required: "Please fill out this field",
  min: "Value must be greater than or equal to ${min}",
  max: "Value must be less than or equal to ${max}",
  url: "Please enter a URL",
  email: "Please include a valid e-mail address",
  pattern: "Please match the requested format",
  maxLength: "Value must have no more than ${maxLength} characters"
};

export default {
  get(key, props) {
    return template(contents[key], props);
  },

  set(newContents) {
    contents = Object.assign(contents, newContents);
  }
};
