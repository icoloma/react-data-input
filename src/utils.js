function isNullOrUndefined(value) {
  return typeof value === "undefined" || value === null;
}
function isBlank(str) {
  return !str || /^\s*$/.test(str);
}

function isFalse(value) {
  return value === false;
}

export { isNullOrUndefined, isBlank, isFalse };
