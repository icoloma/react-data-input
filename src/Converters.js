import { isNullOrUndefined, isBlank, passThrough } from "./utils";

const Converters = {
  text: {
    // transform from String representation to object.
    toObject: passThrough,

    // transform from object to String. Null and undefined values are converted to an empty String
    toString: function(value) {
      return isNullOrUndefined(value) ? "" : value;
    }
  },

  url: {
    // transform from String representation to object.
    toObject: function(value) {
      return isBlank(value) ? undefined : value;
    },

    toString: function(value) {
      return Converters.text.toString(value);
    }
  },

  number: {
    toObject: function(value, attrs) {
      return isBlank(value)
        ? undefined
        : attrs.step == 1
        ? parseInt(value)
        : parseFloat(value);
    },

    toString: function(value) {
      return isNullOrUndefined(value) ? "" : "" + value;
    }
  },

  checkbox: {
    toObject: passThrough,
    toString: passThrough
  }
};

export default Converters;
