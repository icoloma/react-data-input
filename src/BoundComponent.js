import React from "react";
import PropTypes from "prop-types";
import isPromise from "is-promise";
import Validators from "./Validators";
import Converters from "./Converters";

// component that will propagate any changes to state[props.name]
export default class BoundComponent extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.onChange = this.onChange.bind(this);
    this.componentWillReceiveProps(props);
  }

  componentWillReceiveProps(props) {
    this.converter =
      props.converter || Converters[props.type] || Converters.text;
    this.state = this.state || {};
    this.state.value = this.getStateObject()[props.name];
  }

  getStateObject() {
    return this.props.state || this.context.form.props.state;
  }

  // get the property from the element after passing it through the Converter
  // the property will be element.value most of the time except for checkboxes and radio buttons
  getElementValue(element) {
    const props = this.props;
    const valueProp = props.type === "checkbox" ? "checked" : "value";
    return this.converter.toObject(element[valueProp], props);
  }

  onChange(e) {
    const { name, onChange, type } = this.props;
    const value = this.getElementValue(e.target);
    this.setState({ value, errorMessage: undefined });
    this.getStateObject()[name] = value;
    onChange && onChange(e);
    if (type === "radio") {
      this.context.radioGroup.onChange();
    }
  }

  componentDidMount() {
    this.context.form.addValidationComponent(this);
  }

  componentWillUnmount() {
    this.context.form.removeValidationComponent(this);
  }

  // launch the custom validation from props.validator
  customValidation(value) {
    const props = this.props;
    return !props.validator ? undefined : props.validator(value, props);
  }

  // @return a Promise with the validation result. The result is the error message to display
  validate() {
    const allProps = this.props;
    const validationProps = Validators.filterValidationProps(allProps);
    const value = this.state.value;
    let messageOrPromise = undefined;

    // validator for each property
    Object.keys(validationProps).find(prop => {
      const validator =
        prop === "type"
          ? Validators[allProps.type]
          : Validators[allProps.type + "." + prop] || Validators[prop];
      if (validator) {
        return (messageOrPromise = validator(value, allProps));
      }
    });

    // custom validator specified by the user
    messageOrPromise = messageOrPromise || this.customValidation(value);

    // return promise in every case, and change state to reflect the change
    const promise = isPromise(messageOrPromise)
      ? messageOrPromise
      : Promise.resolve(messageOrPromise);
    return promise.then(errorMessage => {
      this.setState({ errorMessage });
      return errorMessage;
    });
  }

  getNestedElementProps() {
    const value = this.converter.toString(this.state.value);
    const { state, validator, converter, ...props } = this.props;
    const errorMessageId = this.getErrorMessageId();
    if (props.type === "checkbox") {
      props.checked = value;
    } else if (props.type === "radio") {
      props.checked = props.value == value;
    } else {
      props.value = value;
    }
    if (errorMessageId) {
      props["aria-invalid"] = true;
      props["aria-describedBy"] = errorMessageId;
    }
    props.onChange = this.onChange;
    return props;
  }

  getErrorMessageId() {
    return this.state.errorMessage
      ? `${this.props.name || this.props.id}_error`
      : undefined;
  }

  render() {
    const { errorMessage } = this.state;
    const errorMessageElement = errorMessage ? (
      <div className="input-error" id={this.getErrorMessageId()} role="alert">
        {errorMessage}
      </div>
    ) : (
      undefined
    );
    const props = this.getNestedElementProps();
    const element = React.createElement(
      this.getNestedElementClass(),
      props,
      props.children
    );
    return (
      <div className={"input-wrapper " + (props.type || "")}>
        {element}
        {errorMessageElement}
      </div>
    );
  }
}

BoundComponent.propTypes = {
  // name of the input field. Will be used to set the value into the state object
  name: PropTypes.string.isRequired,

  // the state from the parent Form can be overriden at the element level
  state: PropTypes.object
};

BoundComponent.contextTypes = {
  form: PropTypes.object.isRequired,
  radioGroup: PropTypes.object
};
