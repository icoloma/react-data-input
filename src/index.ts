import { Converters } from "./core/Converters";
import { ValidationActions } from "./core/ValidationActions";
import { Input, TextArea, Select } from "./components/InputElements";
import { Form, FieldSet } from "./components/Form";
import { Button } from "./components/Button";

export {
  Form,
  FieldSet,
  Button,
  Input,
  TextArea,
  Select,
  Converters,
  ValidationActions,
};
