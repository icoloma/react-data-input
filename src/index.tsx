import { Converters } from "./core/Converters";
import { ValidationActions } from "./core/ValidationActions";
import { Input, TextArea, Select } from "./components/Elements";
import { Form } from "./components/Form";
import { Button } from "./components/Button";

export { Form, Button, Input, TextArea, Select, Converters, ValidationActions };
