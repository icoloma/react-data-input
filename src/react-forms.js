import Converters from './Converters';
import Messages from './Messages';
import Validators from './Validators';
import { Input, TextArea, Select } from './Elements';
import Form from './Form';
import RadioGroup from './RadioGroup';

export { Form, RadioGroup, Input, TextArea, Select, Converters, Messages, Validators };
