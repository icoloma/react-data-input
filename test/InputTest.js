import React from "react";
import { Form, Input, RadioGroup } from "../src/react-data-input";
import renderer from "react-test-renderer";

function noop() {}
describe("Input", function() {
  function mount(input, state) {
    return renderer.create(
      <Form onSubmit={noop} state={state}>
        {input}
      </Form>
    );
  }

  function getInputElement(form) {
    const inputElement = form.root.find(el => el.type == "input");
    expect(inputElement.type).toBe("input");
    return inputElement;
  }

  it("should render correctly", function() {
    const state = { name: "foobar" };
    const form = mount(
      <Input type="text" className="foo" required name="name" />,
      state
    );
    // input has class 'foo'
    // input is required
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("should render a validation error", function() {
    const state = {};
    const form = mount(<Input type="text" required name="name" />, state);
    return form
      .getInstance()
      .validationComponents[0].validate()
      .then(() => {
        // error message was rendered
        // error message is accessible: role=alert, aria-invalid, aria-describedby
        expect(form.toJSON()).toMatchSnapshot();
      });
  });

  it("should update int on change", function() {
    const state = { age: 20, color: "red" };
    const callback = jest.fn(noop);
    const form = mount(
      <Input
        type="number"
        name="age"
        state={state}
        step="1"
        onChange={callback}
      />,
      state
    );
    const input = getInputElement(form);
    input.props.onChange({
      target: { value: "32.2" }
    });

    expect(state).toMatchObject({
      age: 32,
      color: "red"
    });
    expect(callback).toHaveBeenCalledTimes(1);
  });

  it("should update float on change", function() {
    const state = { age: 20 };
    const form = mount(<Input type="number" name="age" step="0.01" />, state);
    const input = getInputElement(form);
    input.props.onChange({
      target: { value: "32.2" }
    });
    expect(state.age).toBe(32.2);
  });

  it("should use checked with type=checkbox", function() {
    const state = { subscribed: true };
    const form = mount(<Input type="checkbox" name="subscribed" />, state);
    // input[checked] is set
    // input[value] is not set
    expect(form.toJSON()).toMatchSnapshot();
  });

  it("should not throw required error with not required, unchecked checkboxes", function() {
    const state = { subscribed: false };
    const form = mount(
      <Input type="checkbox" name="subscribed" required={false} />,
      state
    );
    return form
      .getInstance()
      .validationComponents[0].validate()
      .then(() => {
        // no error component
        expect(form.toJSON()).toMatchSnapshot();
      });
  });

  it("should throw required error with required, unchecked checkboxes", function() {
    const state = { subscribed: false };
    const form = mount(
      <Input type="checkbox" name="subscribed" required={true} />,
      state
    );
    return form
      .getInstance()
      .validationComponents[0].validate()
      .then(() => {
        // shows error component
        expect(form.toJSON()).toMatchSnapshot();
      });
  });

  it("should not propagate specific properties to the HTML5 element", function() {
    const form = mount(
      <Input type="text" className="foo" state={{ foo: "foo" }} name="name" />,
      {}
    );
    // state attribute should not be propagated to nested <input>
    expect(getInputElement(form).props.state).toBeUndefined();
  });

  function initAsyncValidator(state) {
    const validator = value =>
      new Promise(resolve => {
        resolve(value === "abc" ? undefined : "Validation failed");
      });
    const form = mount(
      <Input type="text" name="foo" validator={validator} />,
      state
    );
    return form.getInstance();
  }

  it("custom validator with Promise expected to fail", function() {
    const state = { foo: "xyz" };
    const form = initAsyncValidator(state);
    return form.validate().then(result => {
      expect(result).toBe(false);
    });
  });

  it("custom validator with Promise expected to pass", function() {
    const state = { foo: "abc" };
    const form = initAsyncValidator(state);
    return form.validate().then(result => {
      expect(result).toBe(true);
    });
  });

  it("should render type=radio", function() {
    const state = { gender: "male" };
    const form = mount(
      <Form onSubmit={noop} state={state}>
        <RadioGroup>
          <div>
            <Input type="radio" name="gender" value="male" />,
            <Input type="radio" name="gender" value="female" />
          </div>
        </RadioGroup>
      </Form>,
      state
    );
    // male selected, female unselected
    expect(form.toJSON()).toMatchSnapshot(
      "radio: male selected, female unselected"
    );

    // trigger change in input2
    const input2 = form.root.findAll(el => el.type == "input")[1];
    input2.props.onChange({
      target: { value: "female", checked: true }
    });
    expect(form.toJSON()).toMatchSnapshot(
      "radio: female selected, male unselected"
    );
  });

  it("should not throw required error with empty, optional emails", function() {
    const form = mount(<Input type="email" name="email" />, {
      email: ""
    });
    return form
      .getInstance()
      .validationComponents[0].validate()
      .then(() => {
        // no input error
        expect(form.toJSON()).toMatchSnapshot();
      });
  });

  it("should throw required error with required emails", function() {
    const form = mount(<Input type="email" name="email" required />, {
      email: ""
    });
    return form
      .getInstance()
      .validationComponents[0].validate()
      .then(() => {
        // should show input error
        expect(form.toJSON()).toMatchSnapshot();
      });
  });

  it("should add aria-describedby if there are validation errors", function() {
    const form = mount(
      <Input type="text" name="name" required aria-describedby="foobar" />,
      {
        name: ""
      }
    );
    return form
      .getInstance()
      .validationComponents[0].validate()
      .then(() => {
        // should prepend the error message to the existing aria-describedby field
        expect(form.toJSON()).toMatchSnapshot();
      });
  });

  it("should render validation errors when the form value is not an e-mail", function() {
    const form = mount(<Input type="email" name="email" />, {
      email: "foo"
    });
    return form
      .getInstance()
      .validationComponents[0].validate()
      .then(() => {
        // should show input error
        expect(form.toJSON()).toMatchSnapshot();
      });
  });
});
