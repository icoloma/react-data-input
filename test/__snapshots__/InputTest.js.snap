// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should add aria-describedby if there are validation errors 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="input-wrapper text"
  >
    <input
      aria-describedby="name_error foobar"
      aria-invalid={true}
      name="name"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    <div
      className="input-error"
      id="name_error"
      role="alert"
    >
      Please fill out this field
    </div>
  </div>
</form>
`;

exports[`Input should not throw required error with empty, optional emails 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="input-wrapper email"
  >
    <input
      name="email"
      onChange={[Function]}
      type="email"
      value=""
    />
  </div>
</form>
`;

exports[`Input should not throw required error with not required, unchecked checkboxes 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="input-wrapper checkbox"
  >
    <input
      checked={false}
      name="subscribed"
      onChange={[Function]}
      required={false}
      type="checkbox"
    />
  </div>
</form>
`;

exports[`Input should render a validation error 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="input-wrapper text"
  >
    <input
      aria-describedby="name_error"
      aria-invalid={true}
      name="name"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    <div
      className="input-error"
      id="name_error"
      role="alert"
    >
      Please fill out this field
    </div>
  </div>
</form>
`;

exports[`Input should render correctly 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="input-wrapper text"
  >
    <input
      className="foo"
      name="name"
      onChange={[Function]}
      required={true}
      type="text"
      value="foobar"
    />
  </div>
</form>
`;

exports[`Input should render type=radio: radio: female selected, male unselected 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <div>
      <div>
        <div
          className="input-wrapper radio"
        >
          <input
            checked={false}
            name="gender"
            onChange={[Function]}
            type="radio"
            value="male"
          />
        </div>
        ,
        <div
          className="input-wrapper radio"
        >
          <input
            checked={true}
            name="gender"
            onChange={[Function]}
            type="radio"
            value="female"
          />
        </div>
      </div>
    </div>
  </form>
</form>
`;

exports[`Input should render type=radio: radio: male selected, female unselected 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <div>
      <div>
        <div
          className="input-wrapper radio"
        >
          <input
            checked={true}
            name="gender"
            onChange={[Function]}
            type="radio"
            value="male"
          />
        </div>
        ,
        <div
          className="input-wrapper radio"
        >
          <input
            checked={false}
            name="gender"
            onChange={[Function]}
            type="radio"
            value="female"
          />
        </div>
      </div>
    </div>
  </form>
</form>
`;

exports[`Input should render validation errors when the form value is not an e-mail 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="input-wrapper email"
  >
    <input
      aria-describedby="email_error"
      aria-invalid={true}
      name="email"
      onChange={[Function]}
      type="email"
      value="foo"
    />
    <div
      className="input-error"
      id="email_error"
      role="alert"
    >
      Please include a valid e-mail address
    </div>
  </div>
</form>
`;

exports[`Input should throw required error with required emails 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="input-wrapper email"
  >
    <input
      aria-describedby="email_error"
      aria-invalid={true}
      name="email"
      onChange={[Function]}
      required={true}
      type="email"
      value=""
    />
    <div
      className="input-error"
      id="email_error"
      role="alert"
    >
      Please fill out this field
    </div>
  </div>
</form>
`;

exports[`Input should throw required error with required, unchecked checkboxes 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="input-wrapper checkbox"
  >
    <input
      aria-describedby="subscribed_error"
      aria-invalid={true}
      checked={false}
      name="subscribed"
      onChange={[Function]}
      required={true}
      type="checkbox"
    />
    <div
      className="input-error"
      id="subscribed_error"
      role="alert"
    >
      Please fill out this field
    </div>
  </div>
</form>
`;

exports[`Input should use checked with type=checkbox 1`] = `
<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="input-wrapper checkbox"
  >
    <input
      checked={true}
      name="subscribed"
      onChange={[Function]}
      type="checkbox"
    />
  </div>
</form>
`;
